### PR TITLE
Changes for app backup/restore to work with multiple drivers

### PR DESCRIPTION
**Pull request by disrani-px**
**What type of PR is this?**
Feature

**What this PR does / why we need it**:
* When taking a backup we will now look through PVCs in the namespace for all
    supported drivers and try to create a backup for all of them
* The driver name will be stored in the backup object so that we know which
    driver to use when doing the restore
* This also allows us to create a backup with multiple volume drivers in the
    same backup
* Added an options field in the application backup CRD which can be used to
    store info like projectID, zone, etc for cloud drives
* Added implementation for gce driver for application backup and restore
* Updated startup params to start application manager controllers even if no driver is specified
   

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:
```release-note
- Added support to backup and restore PVCs backed by Google Cloud Engine (GCE) Persistent Disks (PD)
```

**Does this change need to be cherry-picked to a release branch?**:

No

### CODE REVIEW

File: drivers/volume/gcp/gcp.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+	volumeInfos := make([]*storkapi.ApplicationRestoreVolumeInfo, 0)
+	for _, backupVolumeInfo := range volumeBackupInfos {
+		volumeInfo := &storkapi.ApplicationRestoreVolumeInfo{}
+		volumeInfo.PersistentVolumeClaim = backupVolumeInfo.PersistentVolumeClaim
+		volumeInfo.SourceNamespace = backupVolumeInfo.Namespace
+		volumeInfo.SourceVolume = backupVolumeInfo.Volume
+		volumeInfo.RestoreVolume = g.generatePVName()
+		volumeInfo.DriverName = driverName
+		volumeInfos = append(volumeInfos, volumeInfo)
```
Comment (pault84) [inline drivers/volume/gcp/gcp.go]: NIT: could do the assignment in the append to have it be a one-liner.

File: drivers/volume/mock/mock.go [hunk 2]
```diff
@@ -251,6 +251,11 @@ func (m *Driver) OwnsPVC(pvc *v1.PersistentVolumeClaim) bool {
 	return true
 }
 
+// OwnsPV returns true because it owns all PVC created by tests
+func (m *Driver) OwnsPV(pv *v1.PersistentVolume) bool {
```
Comment (pault84) [inline drivers/volume/mock/mock.go]: Not sure why we have this function if it always returns true? will this change in the future?

File: cmd/stork/stork.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ import (
 	"syscall"
 
 	"github.com/libopenstorage/stork/drivers/volume"
+	_ "github.com/libopenstorage/stork/drivers/volume/gcp"
```
[the remark this answers is not in the document]
Comment (disrani-px) [reply cmd/stork/stork.go]: Yes, it is needed to register the gcp driver

File: cmd/stork/stork.go [hunk 4]
```diff
@@ -179,14 +181,16 @@ func run(c *cli.Context) {
 	eventBroadcaster.StartRecordingToSink(&core_v1.EventSinkImpl{Interface: core_v1.New(k8sClient.CoreV1().RESTClient()).Events("")})
 	recorder := eventBroadcaster.NewRecorder(legacyscheme.Scheme, api_v1.EventSource{Component: eventComponentName})
 
-	if c.Bool("extender") {
-		ext = &extender.Extender{
-			Driver:   d,
-			Recorder: recorder,
-		}
+	if driverName != "" {
```
[the remark this answers is not in the document]
Comment (disrani-px) [reply cmd/stork/stork.go]: Done

File: drivers/volume/gcp/gcp.go [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+	"google.golang.org/api/option"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	k8shelper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+)
+
+const (
+	// driverName is the name of the gcp driver implementation
+	driverName = "gce"
+	// provisioner names for portworx volumes
```
[the remark this answers is not in the document]
Comment (disrani-px) [reply drivers/volume/gcp/gcp.go]: Fixed

File: drivers/volume/gcp/gcp.go [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+	provisioner := ""
+	// Check for the provisioner in the PVC annotation. If not populated
+	// try getting the provisioner from the Storage class.
+	if val, ok := pvc.Annotations[pvcProvisionerAnnotation]; ok {
+		provisioner = val
+	} else {
+		storageClassName := k8shelper.GetPersistentVolumeClaimClass(pvc)
+		if storageClassName != "" {
+			storageClass, err := k8s.Instance().GetStorageClass(storageClassName)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (disrani-px) [reply drivers/volume/gcp/gcp.go]: This is to print the warning if the storageclass is not found. We still try to get the provisioner from the PV if this fails

File: drivers/volume/gcp/gcp.go (hunk 1 above)
Comment (disrani-px) [reply drivers/volume/gcp/gcp.go]: Need to use `volumeInfo` below, so can't do that. Moved the assignments inside the init of the object

File: drivers/volume/gcp/gcp.go [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+func (g *gcp) GetVolumeClaimTemplates([]v1.PersistentVolumeClaim) (
+	[]v1.PersistentVolumeClaim, error) {
+	return nil, &errors.ErrNotSupported{}
+}
+
+func init() {
+	g := &gcp{}
+	err := g.Init(nil)
+	if err != nil {
+		logrus.Errorf("Error init'ing gcp driver")
```
[the remark this answers is not in the document]
Comment (disrani-px) [reply drivers/volume/gcp/gcp.go]: Fixed

File: drivers/volume/mock/mock.go (hunk 2 above)
Comment (disrani-px) [reply drivers/volume/mock/mock.go]: This is a mock driver which is only used for testing. This is a part of the driver interface, each driver returns the correct value. For unit tests we use the mock driver which always return true

**Pull request merged**